### PR TITLE
Failing volume create in case of wrong/incompatible VSAN policy

### PR DESCRIPTION
**Pull request by pshahzeb**
1. In case of wrong/incompatible policy, error is returned which is
   displayed on the client. Changing signature of set_policy to return
   the error thrown by VSAN
2. cleanVMDK to delete the vmdk file in cases when metafile hasn't been
   created but the vmdk is; and the vmdk file needs to be deleted
   Invoked as a part of removeVMDK procedure and
   cases requiring deletion of vmdk file only (when meta file
   hasn't been generated)
   eg: Unsuccesful attempt to apply vsan policy and when failed
   to create metadata for vmdk_path

Testing:
Added test scenario of impossible policy so that createVMDK fails
Add this in VmdkCreateRemoveTestCase

Manual Testing:
```
[root@sc-rdops-vm07-dhcp-208-78:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py policy create --name bad_policy --content '(("proportionalCapacity" i0)("hostFailuresToTolerate" i3))'
Successfully created policy: bad_policy

root@photon-HqYoHLUks [ ~ ]# docker volume create --driver=vmdk --name=vol_test -o vsan-policy-name=bad_policy
Error response from daemon: create vol_test: VolumeDriver.Create: Failed to create volume vol_test: Failed to set object policy : Out of resources 1835026.
object setPolicy error: Failure


[root@sc-rdops-vm07-dhcp-208-78:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py policy create --name wrong_string_policy --content '(("proportionalCapacity" i0)("hostFailuresToTolerate" i0)'
Successfully created policy: wrong_string_policy

root@photon-HqYoHLUks [ ~ ]# docker volume create --driver=vmdk --name=vol_test2 -o vsan-policy-name=wrong_string_policy
Error response from daemon: create vol_test2: VolumeDriver.Create: Failed to create volume vol_test2: error in line 2: syntax error
 VSAN_SPBMToVsanPolicy: Failed to convert SPBM string to VsanExpr: Not ready
Failed to set object policy : Input/output error 327684.
object setPolicy error: Failure
```

First output is the case where the policy contents are valid but it is incompatible. Second case where the contents of the policy itself are wrong.

Resolves: #874

### CODE REVIEW

File: esx_service/vmdk_ops.py [hunk 1]
```diff
@@ -201,22 +201,21 @@ def createVMDK(vmdk_path, vm_name, vol_name, opts={}, vm_uuid=None, tenant_uuid=
...
     if kv.VSAN_POLICY_NAME in opts:
-        if not vsan_policy.set_policy_by_name(vmdk_path, opts[kv.VSAN_POLICY_NAME]):
-            # Drop the failed option
-            # A warning is being logged in the called functions
-            del opts[kv.VSAN_POLICY_NAME]
+        set_err = set_policy_to_vmdk(vmdk_path=vmdk_path,
+                            opts=opts,
+                            vol_name=vol_name)
+        if set_err:
+            return set_err
```
Comment (govint) [inline esx_service/vmdk_ops.py]: Shouldn't the VMDK be cleaned and better removed in all cases where error is returned vs. keeping the VMDK around.

Comment (pshahzeb) [inline esx_service/vmdk_ops.py]: VMDK is removed.  set_policy_to_vmdk() makes call to cleanVMDK() which cleans VMDK in cases where error has occurred. eg: VSAN couldn't set the policy.

Comment (govint) [inline esx_service/vmdk_ops.py]: Ok, didn't check set_policy_to_vmdk(), possibly add a comment here that the set call will remove the VMDK on erro.

Comment (pshahzeb) [inline esx_service/vmdk_ops.py]: Sure.

File: esx_service/vmdk_ops.py [hunk 2]
```diff
@@ -402,6 +403,28 @@ def validate_vsan_policy_name(policy_name, vmdk_path):
...
+        # If policy is incompatible/wrong, return the error and delete the vmdk_path
+        msg = ("Failed to create volume %s: %s" % (vol_name, out))
+        logging.warning(msg)
+        error_info = err(msg)
+        clean_err = cleanVMDK(vmdk_path=vmdk_path,
+                            vol_name=vol_name)
+
+        if clean_err:
+            error_info = error_info + clean_err
+
```
Comment (govint) [inline esx_service/vmdk_ops.py]: Could log the cleanVMDK() error as well.

File: esx_service/vmdk_ops.py [hunk 3]
```diff
@@ -521,7 +541,27 @@ def removeVMDK(vmdk_path, vol_name=None, vm_name=None, tenant_uuid=None, datasto
                 retry_count += 1
```
Comment (govint) [inline esx_service/vmdk_ops.py]: cloneVMDK in log.

File: esx_service/vmdk_ops.py [hunk 4]
```diff
@@ -489,18 +512,15 @@ def vol_info(vol_meta, vol_size_info, datastore):
...
+def cleanVMDK(vmdk_path, vol_name=None):
+    """
+    Delete the vmdk file. Retry if the attempt fails
+    Invoked as a part of removeVMDK procedure and
+    cases requiring deletion of vmdk file only (when meta file
+    hasn't been generated)
+    eg: Unsuccesful attempt to apply vsan policy and when failed
+    to create metadata for vmdk_path
+    """
     # Form datastore path from vmdk_path
```
Comment (govint) [inline esx_service/vmdk_ops.py]: Could include an info log for the VMDK being removed.

File: esx_service/vmdk_ops.py [hunk 5]
```diff
@@ -521,7 +541,27 @@ def removeVMDK(vmdk_path, vol_name=None, vm_name=None, tenant_uuid=None, datasto
...
+            return err("Failed to remove volume {0}, in use by VM uuid = {1}.".format(
+                vmdk_path, kv_uuid))
+
+    # Cleaning .vmdk file
+    clean_err = cleanVMDK(vmdk_path, vol_name)
+
+    if clean_err:
+        return clean_err
+
+    # clean succeed, remove infomation of this volume from volumes table
```
Comment (govint) [inline esx_service/vmdk_ops.py]: Nit - "Clean succeeded...."

File: esx_service/vmdk_ops.py [hunk 6]
```diff
@@ -489,18 +512,15 @@ def vol_info(vol_meta, vol_size_info, datastore):
...
-
+def cleanVMDK(vmdk_path, vol_name=None):
+    """
+    Delete the vmdk file. Retry if the attempt fails
+    Invoked as a part of removeVMDK procedure and
+    cases requiring deletion of vmdk file only (when meta file
+    hasn't been generated)
+    eg: Unsuccesful attempt to apply vsan policy and when failed
+    to create metadata for vmdk_path
+    """
```
Comment (govint) [inline esx_service/vmdk_ops.py]: Actually, removal of the volume will automatically result in the meta-data file also getting removed.. Why is this function needed?

Comment (pshahzeb) [inline esx_service/vmdk_ops.py]: Removal of a volume gracefully will do it. But other cases are when the vmdk file was created but the meta file wasn't. Like unsuccessful attempt to apply vsan policy and when kv.create() method itself fail. In such cases, we just need to delete the vmdk file.

File: esx_service/vmdk_ops.py [hunk 7]
```diff
@@ -201,22 +201,21 @@ def createVMDK(vmdk_path, vm_name, vol_name, opts={}, vm_uuid=None, tenant_uuid=
...
-                                vm_name=vm_name,
-                                tenant_uuid=tenant_uuid,
-                                datastore=datastore)
-        if remove_err:
-            error_info = error_info + remove_err
+        clean_err = cleanVMDK(vmdk_path=vmdk_path,
+                            vol_name=vol_name)
+
+        if clean_err:
+            error_info = error_info + clean_err
```
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Please log the error here.

File: esx_service/vmdk_ops.py [hunk 8]
```diff
@@ -521,7 +541,27 @@ def removeVMDK(vmdk_path, vol_name=None, vm_name=None, tenant_uuid=None, datasto
                 retry_count += 1
                 time.sleep(vmdk_utils.VMDK_RETRY_SLEEP)
 
-    # remove succeed, remove infomation of this volume from volumes table
+    return None
+
+# Return error, or None for OK
+def removeVMDK(vmdk_path, vol_name=None, vm_name=None, tenant_uuid=None, datastore=None):
```
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Please add docstring for  function "removeVMDK" to say that removeVMDK handle the normal volume remove (remove .vmdk file and clean the metadata).

File: esx_service/vmdk_ops_test.py [hunk 9]
```diff
@@ -114,16 +114,23 @@ def setUp(self):
...
 
         self.vmdk = None
-        for n in self.policy_names:
-            vsan_policy.delete(n)
+
+        vsan_policy.delete('good')
+        vsan_policy.delete('impossible')
 
     def testCreateDelete(self):
         err = vmdk_ops.createVMDK(vm_name=self.vm_name,
```
Comment (lipingxue) [inline esx_service/vmdk_ops_test.py]: Please check if existing automated test already covered the second manual test part, if not, please add it .

File: esx_service/vmdk_ops.py [hunk 10]
```diff
@@ -419,6 +423,7 @@ def set_policy_to_vmdk(vmdk_path, opts, vol_name=None):
                             vol_name=vol_name)
 
         if clean_err:
+            logging.warning("Failed to clean %s file", vmdk_path)
```
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Better to log the error_info which will provide more information about why it fails.

File: esx_service/utils/vsan_info_test.py [hunk 11]
```diff
@@ -84,9 +84,8 @@ def test_policy(self):
         # different content:
         notsame_policy = \
             '(("hostFailuresToTolerate" i0) ("forceProvisioning" i0))'
-        self.assertTrue(
-            vsan_info.set_policy(self.VMDK_PATH, policy_string),
-            "failed to set")
+        err = vsan_info.set_policy(self.VMDK_PATH, policy_string)
```
Comment (shuklanirdesh82) [inline esx_service/utils/vsan_info_test.py]: minor: I guess it is OK to have it like that

File: esx_service/vmdk_ops.py [hunk 12]
```diff
@@ -305,16 +306,20 @@ def cloneVMDK(vm_name, vmdk_path, opts={}, vm_uuid=None, vm_datastore=None):
         except vim.fault.VimFault as ex:
             return err("Failed to clone volume: {0}".format(ex.msg))
 
+    vol_name = vmdk_utils.strip_vmdk_extension(src_vmdk_path.split("/")[-1])
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops.py]: minor: please make sure the formatting is aligned correctly.

Comment (pshahzeb) [inline esx_service/vmdk_ops.py]: Oh this statement is indented. Meant to be outside the loop which is above it

File: esx_service/utils/vsan_info_test.py (hunk 11 above)
Comment (pshahzeb) [inline esx_service/utils/vsan_info_test.py]: The signature of set_policy changed, hence had to change the check.

**Pull request merged**